core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/Avg.java

```diff
@@ -38,32 +43,45 @@ public Avg(String fieldName, Schema fieldSchema) {
         "Cannot compute avg on field %s because its type %s is not numeric", fieldName, fieldType));
     }
 
+    // the avg is null only if the field value is always null
     outputSchema = isNullable ? Schema.nullableOf(Schema.of(Schema.Type.DOUBLE)) : Schema.of(Schema.Type.DOUBLE);
   }
 
   @Override
-  public void beginFunction() {
-    avg = 0d;
-    count = 0d;
+  public void mergeValue(Table<String, String, Object> agg, StructuredRecord record) {
+    Object val = record.get(fieldName);
+    if (val == null) {
+      return;
+    }
+    computeAvg(agg, 1L, (Number) val);
   }
 
   @Override
-  public void operateOn(StructuredRecord record) {
-    Object val = record.get(fieldName);
+  public void mergeAggregates(Table<String, String, Object> agg1, Table<String, String, Object> agg2) {
+    Object val = agg2.get(fieldName, AGG_KEY);
     if (val == null) {
       return;
     }
-    count++;
-    avg = avg + (((Number) val).doubleValue() - avg) / count;
+    Long agg2Count = (Long) agg2.get(fieldName, AGG_COUNT_KEY);
+    computeAvg(agg1, agg2Count, (Number) val);
   }
 
-  @Override
-  public Double getAggregate() {
-    if (count == 0) {
-      // only happens if the field value was always null
-      return null;
+  private void computeAvg(Table<String, String, Object> table, long deltaCount, Number oldAvg) {
+    if (!table.contains(fieldName, AGG_KEY)) {
```

Comment:
ok so it looks like Table is used to prevent conflicts between fields in the group key and intermediate data?

It seems like it would be better to have a separate class that contains the group key as a field along with any additional fields that are required.